### PR TITLE
Fix: avoid using regex match, possibly containing alternation, as a key condition.

### LINKED ISSUE #53222: Using match with `^$` as anchors on a primary key column returns incorrect results
**Describe what's wrong**

When running a query that applies a regex expression to the primary key column and uses `^$` as anchors on the first element, Clickhouse incorrectly returns only a portion of the expected results. 

**Does it reproduce on recent release?**

Running release `23.3.8.21`.

**How to reproduce**
Create table and insert data into different parts. Note that this bug appears to apply to only the primary key condition as inserting both data items into same part will return the correct result.
```
CREATE TABLE example(
    "time" Int64 CODEC(ZSTD(1)),
    "svc" LowCardinality(String) CODEC(ZSTD(1)),
    "title" String CODEC(ZSTD(1)),
 ) ENGINE = MergeTree
 PARTITION BY intDiv("time", 1000)
 ORDER BY ("svc", "time");
 
INSERT INTO example(*) VALUES(toInt64(4500), 'first', 'blah blah')
INSERT INTO example(*) VALUES(toInt64(3500), 'second', 'blah blah blah')
```

Reading data with `^$` returns only the 1/2 expected rows:
```
SELECT svc, title FROM example WHERE match(svc, '^first$|^second$')
```

Running the query with index explain enabled (`EXPLAIN indexes = 1`) shows the problem right away:
```
Expression ((Projection + Before ORDER BY))
	  ReadFromMergeTree (7_ca_43_de_8_a_195_42_c_7_b_590_7_a_270_a_945433.example)
	  Indexes:
	    MinMax
	      Condition: true
	      Parts: 2/2
	      Granules: 2/2
	    Partition
	      Condition: true
	      Parts: 2/2
	      Granules: 2/2
	    PrimaryKey
	      Keys: 
	        svc
	      Condition: (svc in ['first', 'firsu')) <-- this is the problem
	      Parts: 1/2
	      Granules: 1/2
```
**Additional context**

A simple workaround is to use `\A\z` for anchoring which returns the expected results:
```
SELECT svc, title FROM example WHERE match(svc, '\Afirst\z|\Asecond\z')
```

### PR DESCRIPTION
**Pull request by yakov-olkhovskiy**
### Changelog category (leave one):
- Bug Fix (user-visible misbehavior in an official stable release)

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed bug of match() function (regex) with pattern containing alternation produces incorrect key condition.

closes #53222

### CODE REVIEW

**Comment by robot-ch-test-poll3** [comment]
*This is an automated comment for commit c2d46286e7dd87c6b1fac283c7e299ab3c0537d4 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/54696/c2d46286e7dd87c6b1fac283c7e299ab3c0537d4/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>✅ success</td></tr>
<tbody>
</table>
</details>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>⏳ pending</td></tr>
<tbody>
</table>

File: src/Storages/MergeTree/KeyCondition.cpp [hunk 1]
```diff
@@ -421,7 +421,12 @@ const KeyCondition::AtomMap KeyCondition::atom_map
             if (value.getType() != Field::Types::String)
                 return false;
 
-            String prefix = extractFixedPrefixFromRegularExpression(value.get<const String &>());
+            const String & expression = value.get<const String &>();
+            // This optimization can't process alternation - this would require a comprehensive parsing of regular expression.
```
Comment (rschu1ze) [inline src/Storages/MergeTree/KeyCondition.cpp:425]: I thought a bit 
- whether the check should rather be moved into `extractFixedPrefixFromRegularExpression()` but I think it is okay to not do so
- whether `extractFixedPrefixFromRegularExpression()` could be replaced by the more sophisticated regex parsing in `OptimizedRegularExpression::analyze()`, but again no. `::analyze()` searches the *longest* substring and that's not necessarily the prefix.

Therefore, we can merge the fix. Thanks.

**Review by rschu1ze — APPROVED** [review]

**Comment by den-crane** [comment]
Strange that it's not backported into LTS releases but issue was marked as major

**Comment by rschu1ze** [comment]
True. Let's backport it.

**Pull request merged**